exploration/code-mode-introducer.md

```diff
@@ -0,0 +1,224 @@
+
+Sample pattern:
+```
+{{
+input {$var}
+match {$var}
+when * {{Pattern}}
+}}
+```
+Sample quoted pattern with no declarations or match:
+```
+{{{{Pattern}}}}
+```
+
+Pros:
+- Uses a sigil `{` already present in the syntax
+- No additional escapes
+- Consistent with other parts of the syntax?
+
+Cons:
+- Somewhat verbose
+- Closing portion of the syntax adds no value;
+  could be a source of unintentional syntax errors
+- Messages commonly end with four `}}}}`
+
+### Option B. Use a Sigil
+
+Complex messages start with a special sigil character.
+
+```
+^input {$var}
+match {$var}
+when * {{Pattern}}
+```
+Sample quoted pattern with no declarations or match:
+```
+^{{Pattern}}
+```
+
+Pros:
+- Requires minimum additional typing
+
+Cons:
+- Requires an additional sigil
+- Requires an additional escape for simple pattern start
+- Has no other purpose in the syntax
+
+### Option C. Use a Double Sigil
+
+Like Option B, except the sigil is doubled.
+
+```
+^^input {$var}
+match {$var}
+when * {{Pattern}}
+```
+Sample quoted pattern with no declarations or match:
+```
+^^{{Pattern}}
+```
+
+Pros:
+- Less likely to conflict with a simple pattern
+
+Cons:
+- Requires an additional sigil
+- Requires an additional escape for simple pattern start
+- Has no other purpose in the syntax
+
+### Option D. Sigilized Keywords
+
+Instead of quoting the message, adds a sigil to keywords that
+start statements, that is, `#input`, `#local` and `#match`.
+The keyword `when` might be considered separately.
+
+```
+#input {$var}
+#local $foo = {$bar}
+#match {$var}
+when * {{Pattern}}
+```
+Sample quoted pattern with no declarations or match:
+```
+{{Pattern}}
+```
+
+Pros:
+- Sigil is part of the keyword, not something separate
+- Requires minimum additional typing
+- Adds no characters to messages that consist of only a quoted pattern;
+  that is, quoting the pattern consists only of adding the `{{`/`}}` quotes
+- Maybe makes single-line messages easier to parse visually???
+
+Cons:
+- Requires an additional sigil
+- Requires an additional escape for simple pattern start
+
+> [!Note] Unlike the other options, Option D is presented with `#` as the sigil.
+> That doesn't mean that this should be the sigil.
+> However, `#` was originally chosen for similarity in declarations to `#define` and `#import`
+> in various programming languages.
+> It is kept "as-is" here.
+
+### Option E. Special Sequence
+
+Like Option A except the sequence is closed locally (not at the end of the message).
+The suggested sequence is `{#}` but might be `{}` or `{{}}` also.
+
+```
+{^}input {$var}
+match {$var}
+when * {{Pattern}}
+```
+Sample quoted pattern with no declarations or match:
+```
+{^}{{Pattern}}
+```
+
+Pros:
+- Less likely to conflict with a simple pattern
+- Requires no additional sigil
+- Requires no additional escape
+
+Cons:
+- Has no other purpose in the syntax
+- Looks like something should happen inside it
+- Most additional typing
+  
```

Comment:
```suggestion

### Option F. Preamble

In this option, all declarations are placed in a dedicated block at the beginning of the message.
The preamble is the "front-matter" of the message, containing the message's logic.
`when` clauses are not part of the preamble.

The preamble can be delimited with `{% ... %}`:

    {%input {$var} match {$var}%} when * {{Pattern}}

Alternatively, it can be delimited with a new kind of delimiter, to make it visually distinct from placeholders and patterns:

    [[input {$var} match {$var}]] when * {{Pattern}}
    
We could also consider dropping the `when` keywords:

    [[input {$var} match {$var}]] * {{Pattern}}

Pros:
- Provides a clear conceptual distinction between declarations and variants.
- Visually, all code is grouped together.
- Unnests variant patterns. 

Cons:
- If `[[ ... ]]` is used to delimit the preamble, it will require `[[` to be escaped in patterns.

```